### ISSUE TITLE
Nomenclature for Equipment and Devices

### ISSUE THREAD

**Issue by daanpape** [PoseiDAT/schemas#34]
At DBMatic we are currently connecting more and more advanced equipment to the concentrator. The more complex things are getting the more I am convinced that `Equipment` and `Device` are not the correct terms for what we are using them. Take for example a ship engine, currently this would be described in the following way:

```
equipment: main engine
   |
  +- device: cylinder_1
       |
      +- Measurement type: Temperature
  +- device: cylinder_2
       |
      +- Measurement type: Temperature
  +- device: cylinder_3
       |
      +- Measurement type: Temperature
  +- device: cylinder_4
       |
      +- Measurement type: Temperature
  +- device: cylinder_5
       |
      +- Measurement type: Temperature
  +- device: cylinder_6
       |
      +- Measurement type: Temperature
  +- device: cooling_water_in
       |
      +- Measurement type: Temperature
      +- Measurement type: Flow
      +- Measurement type: Pressure
  +- device: cooling_water_out
       |
      +- Measurement type: Temperature
      +- Measurement type: Flow
      +- Measurement type: Pressure
  +- device: oil_carter
       |
      +- Measurement type: Temperature
      +- Measurement type: Volume
      +- Measurement type: Level
```

The above doesn't feel very good to me and therefore I would like to propose to change the term `device` to `component` as this is more generic. This would change the above to the following structure:

```
equipment: main engine
   |
  +- component: cylinder_1
       |
      +- Measurement type: Temperature
  +- component: cylinder_2
       |
      +- Measurement type: Temperature
  +- component: cylinder_3
       |
      +- Measurement type: Temperature
  +- component: cylinder_4
       |
      +- Measurement type: Temperature
  +- component: cylinder_5
       |
      +- Measurement type: Temperature
  +- component: cylinder_6
       |
      +- Measurement type: Temperature
  +- component: cooling_water_in
       |
      +- Measurement type: Temperature
      +- Measurement type: Flow
      +- Measurement type: Pressure
  +- component: cooling_water_out
       |
      +- Measurement type: Temperature
      +- Measurement type: Flow
      +- Measurement type: Pressure
  +- component: oil_carter
       |
      +- Measurement type: Temperature
      +- Measurement type: Volume
      +- Measurement type: Level
```

Another problem to tackle is how a PoseiDAT user can know what measurements they can expect from a certain `component/device`, this is not defined in the standard. The only way to know this is to request measurements from a certain `component/device` and see what types are returned.

This is not reliable, take for example the `oil_carter` component from the above equipment. It could very well be that the level measurement is sent with another frequency than the temperature measurement. When you then request messages from a time window in which only temperature is transmitted you could falsely assume that level is not measured.

I would therefore like to introduce something called `properties` which can be attached to components in the equipment inventory. A property has a certain type and you can assign a value. This would expand the above equipment to:

```
equipment: main engine
   |
  +- component: cylinder_1
       |
      +- Property type: Volume
      +- Measurement type: Temperature
  +- component: cylinder_2
       |
      +- Property type: Volume
      +- Measurement type: Temperature
  +- component: cylinder_3
       |
      +- Property type: Volume
      +- Measurement type: Temperature
  +- component: cylinder_4
       |
      +- Property type: Volume
      +- Measurement type: Temperature
  +- component: cylinder_5
       |
      +- Property type: Volume
      +- Measurement type: Temperature
  +- component: cylinder_6
       |
      +- Property type: Volume
      +- Measurement type: Temperature
  +- component: cooling_water_in
       |
      +- Property type: min_temperature
      +- Property type: max_temperature
      +- Measurement type: Temperature
      +- Measurement type: Flow
      +- Measurement type: Pressure
  +- component: cooling_water_out
       |
      +- Property type: min_temperature
      +- Property type: max_temperature
      +- Measurement type: Temperature
      +- Measurement type: Flow
      +- Measurement type: Pressure
  +- component: oil_carter
       |
      +- Property type: Volume
      +- Property type: oil_thickness
      +- Property type: min_temperature
      +- Property type: max_temperature
      +- Measurement type: Temperature
      +- Measurement type: Level
```

Before implementing this I would like to have a discussion about the above changes as they are breaking.

Kind regards,
Daan

**Comment by Qwerios** [comment]
I understand the need for the component moniker based on the examples provided. In my head the main question is what is a part of what exactly. To handle the properties you suggested first, I think these should be a part of the equipment list because they will not change and if they do the piece of equipment will likely have been replaced. Being replaced means it gets a new uuid and definition. We opted to use the equipment list to reduce the payload in the measurements. Repeating those properties each time can be avoided. To get the full picture I'd like to see the equipment list to goes along with this example @daanpape 

Right now for equipment we register:

* uuid
* name
* type (ENGINE, AIS, RUDDER, etc)
* supplier (company)
* installer (company)
* maintainer (company)

For a device we register:

* uuid
* name
* type (PUMP, DISPLAY, SCALE, etc)
* brand
* product nr
* serial nr
* supplier (company)
* installer (company)
* maintainer (company)

A device is more like a self contained thing where a component is a smaller part of something else.
For the given example of the main engine it would be strange to refer to the main cylinders as devices.

Going forward we could:

* deprecate devices and replace with components
* add components to a device

Components seem to have less details then devices (no supplier/installer/etc). If this is correct then dropping this extra bloat would be preferable. I would recommend this approach if there is no example of a piece of equipment being comprised of multiple devices. Do components have a product nr and serial that needs to be registered?

To maintain backwards compatibility we could add components to devices. In your main engine example you would end up with:

* equipment: main engine
  * device: cylinders
    * component: cylinder 1
    * component: cylinder 2
    * component: cylinder 3

Measurements should then require a component uuid instead of a device id.
@daanpape do you have any input to help make this decision from the real world of devices?
My preference is leaning towards deprecation as I'd rather break this now while the standard is relatively new.

Your proposed properties is something I have tackled before. We called those sensor classes but for the Poseidat schemas I would call this `measurement-value-type` which could be added to a device/component. We could even add more configuration details like `measurement_interval`. Pushing the device list on each update is vitally important though and I think it should also be possible to request it on-demand at any time.

This is an example list for the measurement value types:

```json
{
  "items": [
    {
      "name": "temperature",
      "unit": "celsius",
      "value_type": "float"
    },
    {
      "name": "pressure",
      "unit": "pascal",
      "value_type": "float"
    },
    {
      "name": "torque",
      "unit": "newton metre",
      "value_type": "float"
    },
    {
      "name": "depth",
      "unit": "meter",
      "value_type": "float"
    },
    {
      "name": "velocity",
      "unit": "meter per second",
      "value_type": "float"
    },
    {
      "name": "location",
      "unit": "dms",
      "value_type": "string"
    },
    {
      "name": "wind-speed",
      "unit": "meter per second",
      "value_type": "float"
    },
    {
      "name": "viscosity",
      "unit": "pascal per second",
      "value_type": "float"
    },
    {
      "name": "flow",
      "unit": "liter per hour",
      "value_type": "float"
    },
    {
      "name": "duration",
      "unit": "seconds",
      "value_type": "integer"
    },
    {
      "name": "power",
      "unit": "watt",
      "value_type": "integer"
    },
    {
      "name": "wind-direction",
      "unit": "degrees 360",
      "value_type": "integer"
    }
  ],
}
```

**Comment by Qwerios** [comment]
After re-reading my reply it might be clearer to indicate the exact amount of sensors we have attached to a component. And that would bring me full circle to sensor classifications. So like this:

* equipement: main engine
  * component: cylinder 1
    * sensor: temperature / celsius / float
  * component: cooling water in
    * sensor: temperature / celsius / float 
    * sensor: flow / liters per h / float
    * sensor: pressure / pascal / float

Making the sensor an actual entity would also leave a place in the equipment list to detail information about the sensor itself. Details of the type, when it was installed, etc

**Comment by daanpape** [comment]
Hello Mark,

Thank you for your view on the subject.

>  In my head the main question is what is a part of what exactly.

My proposition was to change `device` to `component` in the sense that a `component` is part of `equipment` and thus that `equipment` consists out of `components` instead of `devices`. But your last comment is actually even better and allows more profound details. I will update my example below with this new idea.

> Repeating those properties each time can be avoided. To get the full picture I'd like to see the equipment list to goes along with this example @daanpape

We found out is that some `properties` are not manually inserted by an installer but that these come in through data channels such as 'modbus'. I do acknowledge that these could be part of the equipment inventory as these will not change throughout the lifetime of a `component`.

I would therefore propose to create an equipment inventory as follows:

```
{
   "journal_id":"9cc88192-7fc4-4a04-aea6-f25637663525",
   "entry_id":"b10a2f00-f63f-45d5-9458-b25615dbd85e",
   "revision":"2022-01-10T11:02:13.486Z",
   "immutable":false,
   "equipment":[
      {
         "equipment_id":"6bbc79c9-59d5-42dd-a983-21bcc068df73",
         "type":"ENGINE",
         "name":"Main engine",
         "components":[
            {
               "component_id":"fa7dd43d-2d17-495f-ab6d-ff298923bc14",
               "type":"CRANKSHAFT",
               "name":"crankshaft",
               "brand":"ABC",
               "product_no":"6DZC",
               "serial_no":"12345678",
               "sensors":[
                  {
                     "sensor_id":"bcfa3f82-17ec-4f07-9ba2-8d22b844773f",
                     "sensor_type":"SHAFT-SPEED",
                     "measurement_type":"RPM",
                     "name":"primary",
                     "brand":"Bosch",
                     "product_no":"0261210159",
                     "max_value":{
                        
                     },
                     "min_value":{
                        
                     },
                     "measurement_interval":100
                  },
                  {
                     "sensor_id":"bcfa3f82-17ec-4f07-9ba2-8d22b844773f",
                     "sensor_type":"SHAFT-SPEED-SENSOR",
                     "measurement_type":"RPM",
                     "name":"secondary",
                     "brand":"Bosch",
                     "product_no":"0261210159",
                     "max_value":{
                        
                     },
                     "min_value":{
                        
                     },
                     "measurement_interval":100
                  }
               ]
            },
            {
               "component_id":"4ba73817-5f2e-4a32-844d-2d9b1205bbf7",
               "type":"CYLINDER",
               "name":"cylinder_1",
               "brand":"ABC",
               "product_no":"6DZC",
               "serial_no":"12345678",
               "properties":[
                  {
                     "type":"VOLUME-LITER",
                     "value":5.5
                  }
               ],
               "sensors":[
                  {
                     "sensor_id":"1211e660-c16e-4efd-beca-bb0f6b545265",
                     "sensor_type":"TEMPERATURE",
                     "measurement_type":"TEMPERATURE-CELCIUS",
                     "name":"pyrometer_1",
                     "brand":"TiTEC",
                     "product_no":"105391",
                     "max_value":{
                        "value":500
                     },
                     "min_value":{
                        "value":0
                     },
                     "measurement_interval":6000
                  }
               ]
            },
            {
               "component_id":"ca5ec61b-687b-496d-855d-e993c8b6e0e6",
               "type":"CYLINDER",
               "name":"cylinder_2",
               "brand":"ABC",
               "product_no":"6DZC",
               "serial_no":"12345678",
               "properties":[
                  {
                     "type":"VOLUME-LITER",
                     "value":1.2
                  }
               ],
               "sensors":[
                  {
                     "sensor_id":"8316af63-4af7-40e2-8396-31db4f3a3eed",
                     "sensor_type":"TEMPERATURE",
                     "measurement_type":"TEMPERATURE-CELCIUS",
                     "name":"pyrometer_2",
                     "brand":"TiTEC",
                     "product_no":"105391",
                     "max_value":{
                        "value":500
                     },
                     "min_value":{
                        "value":0
                     },
                     "measurement_interval":6000
                  }
               ]
            }
         ]
      }
   ]
}
```
I think deprecation is the way to go for now as the standard is very new and still subject to change.

**Comment by Qwerios** [comment]
Adding `measurement_unit`  (yes on SI) is mandatory I think. Otherwise I think we are in agreement.

I was struggling to see why you would need properties but I see your point. In your example the final measurement would be liters correct? If the raw percentage value would end up on the measure you would always have to back to the equipment list to calculate a usable value. This is further complicated by changing values for `VOLUME-LITER`. Would it make volume measurements always contain 2 values, liters and % of total? Logging this directly would prevent the need for historical lookups in the equipment list.

So to summarise:

* agreed on equipment/comp/sensor and deprecate devices
* add unit to sensor (SI)
* property based sensors might log multiple values

**Comment by daanpape** [comment]
Hello Mark,

Thank you for your insights. I agree on having the unit separate from the measurement type and introduce a new enum `Unit` which is a list of all units supported in PoseiDAT and assign this to a sensor. I would, in that case, also apply this to properties.

This would change the above equipment inventory to:

```
{
    "journal_id": "9cc88192-7fc4-4a04-aea6-f25637663525",
    "entry_id": "b10a2f00-f63f-45d5-9458-b25615dbd85e",
    "revision": "2022-01-10T11:02:13.486Z",
    "immutable": false,
    "equipment": [
        {
            "equipment_id": "6bbc79c9-59d5-42dd-a983-21bcc068df73",
			"type": "ENGINE",
            "name": "Main engine",
            "components": [
                {
                    "component_id": "fa7dd43d-2d17-495f-ab6d-ff298923bc14",
					"type": "CRANKSHAFT",
                    "name": "crankshaft",
                    "brand": "ABC",
                    "product_no": "6DZC",
                    "serial_no": "12345678",
					"sensors": [
						{
							"sensor_id": "bcfa3f82-17ec-4f07-9ba2-8d22b844773f",
							"sensor_type": "SHAFT-SPEED",
							"measurement_type" : "ROTATIONAL-SPEED",
							"unit": "RPM",
							"name": "primary",
							"brand": "Bosch",
							"product_no": "0261210159",
							"measurement_interval": 100
						},
						{
							"sensor_id": "bcfa3f82-17ec-4f07-9ba2-8d22b844773f",
							"sensor_type": "SHAFT-SPEED-SENSOR",
							"measurement_type" : "ROTATIONAL-SPEED",
							"unit": "RPM",
							"name": "secondary",
							"brand": "Bosch",
							"product_no": "0261210159",
							"measurement_interval": 100
						}
					]
                },
				{
                    "component_id": "4ba73817-5f2e-4a32-844d-2d9b1205bbf7",
					"type": "CYLINDER",
                    "name": "cylinder_1",
                    "brand": "ABC",
                    "product_no": "6DZC",
                    "serial_no": "12345678",
					"properties": [
						{
							"type": "VOLUME",
							"unit": "LITER"
							"value": 5.5
						}
					],
					"sensors": [
						{
							"sensor_id": "1211e660-c16e-4efd-beca-bb0f6b545265",
							"sensor_type": "TEMPERATURE",
							"measurement_type" : "TEMPERATURE",
							"unit" : "CELCIUS",
							"name": "pyrometer_1",
							"brand": "TiTEC",
							"product_no": "105391",
							"max_value": {
								"value": 500
							},
							"min_value": {
								"value": 0
							},
							"measurement_interval": 6000
						}
					]
                },
				{
                    "component_id": "ca5ec61b-687b-496d-855d-e993c8b6e0e6",
					"type": "CYLINDER",
                    "name": "cylinder_2",
                    "brand": "ABC",
                    "product_no": "6DZC",
                    "serial_no": "12345678",
					"properties": [
						{
							"type": "VOLUME",
							"unit": "LITER"
							"value": 1.2
						}
					],
					"sensors": [
						{
							"sensor_id": "8316af63-4af7-40e2-8396-31db4f3a3eed",
							"sensor_type": "TEMPERATURE",
							"measurement_type" : "TEMPERATURE",
							"unit" : "CELCIUS",
							"name": "pyrometer_2",
							"brand": "TiTEC",
							"product_no": "105391",
							"max_value": {
								"value": 500
							},
							"min_value": {
								"value": 0
							},
							"measurement_interval": 6000
						}
					]
                }
            ]
        }
    ]
}
```

> Would it make volume measurements always contain 2 values, liters and % of total? Logging this directly would prevent the need for historical lookups in the equipment list.

I think the PoseiDAT standard would have to give the installer/equipment the freedom to choose how to send measurements. The unit will show the user how to interpret the value (absolute or relative). I don't think we need to make volume measurements contain two values. As you said before, the tank volume will only change when a new tank is installed which will cause measurements to come from a new sensor. Only when you would lookup measurements of the old tank you would need to look to the old version of the equipment inventory.

**Comment by Qwerios** [comment]
One small addendum I think `measurement_interval` should be `measurement_interval_ms` to clarify its unit.

**Comment by Qwerios** [comment]
@daanpape is the PR for this still in the works?

**Comment by Qwerios** [comment]
@daanpape its been more then 3 years I'm going to close this due to inactivity. Feel free to open the promised PR if its ready
